Sequential trigger spawning (#698)
cylc/cylc-doc: src/user-guide/writing-workflows/external-triggers.rst

   configuration.

.. _Sequential Xtriggers:

Sequential Xtriggers
---------------------------

Parent-less tasks (which don't depend on other tasks upstream in the graph)
naturally spawn out to the runahead limit. This may cause UI clutter, and

> unnecessary (or even pointless?) xtrigger checking if the xtriggers would only

Reviewer: (Haha, I suggested "unnecessary (or even pointless?)" above, but the intention was we should choose one or the other.)

be satisfied in order.

You can use *sequential* xtriggers to avoid this problem: the next instance
of a task (i.e., at the next cycle point) that depends on a sequential xtrigger
will not be spawned until the previous xtrigger is satisfied. The
``wall_clock`` xtrigger is sequential by default.

A trigger can be set as sequential in any or all of the following ways.

> By setting the workflow wide ``sequential xtriggers`` (defaults
> to ``False``) and/or keyword argument ``sequential`` to ``True``/``False`` in

Reviewer 2: ```suggestion
By setting the workflow-wide :cylc:conf:`flow.cylc[scheduling]sequential xtriggers`
(defaults to ``False``) and/or keyword argument ``sequential`` to ``True``/``False`` in
```

the xtrigger declaration:

.. literalinclude:: ../../workflows/xtrigger/sequential/flow.cylc
   :language: cylc

> And/or via a ``sequential`` keyword argument in the xtrigger function
> definition (see :Ref:`Custom Trigger Functions`) itself:

Reviewer 2: ```suggestion
When implementing a :ref:`custom xtrigger <Custom Trigger Functions>`, you can
set the default for the ``sequential`` keyword argument in the xtrigger function
definition itself:
```

.. code-block:: python

>    my_xtrigger(my_in, my_out, sequential=True)

Reviewer 2: ```suggestion
   def my_xtrigger(my_in, my_out, sequential=True)
```

   configuration.

.. _Sequential Trigger Spawning:

Sequential Trigger Spawning

> ---------------------------

Author: The reason why I didn't use "Xtriggers", and opted for "Triggers", is because all through the wider section they are referred to that way.

> ---------------------------

Reviewer: Ah, OK. Still, just using "triggers" could be misleading - it suggests normal triggers could be sequential too.  Maybe we need to clean up the wider section a bit.

Parent-less tasks (having no graph dependencies on other tasks) naturally spawn
out to the runahead limit. The consequence for workflows with large runahead
limits and/or task count may not only be UI clutter but also needless trigger
checking. An example of this would be ``wall_clock``, where there's clearly

> no point in checking the future if the present trigger is unsatisfied.

Author: Spell check didn't like "Parentless" 

> If their xtriggers will only be satisfied in order anyway

Well, it may cause clutter even if satisfied out of order.. It's really up to the workflow designer if they want to constrain it unnaturally too.

> no point in checking the future if the present trigger is unsatisfied.

Reviewer: True.  In that case, how about:
> This may cause UI clutter, and unnecessary (or even pointless?) xtrigger checking if the xtriggers would only be satisfied in order.

Sequential trigger spawning is a way of altering the default spawning behavior
of these parent-less tasks. This is achieved by setting any number of triggers
as sequential so that associated tasks only spawn their next (cycle point)

> instance on satisfaction of these triggers.

Reviewer: You can use *sequential* xtriggers to avoid this problem: the next instance of a task (i.e., at the next cycle point) that depends on a sequential xtrigger will not be spawned until the previous xtrigger is satisfied. The ``wall_clock`` xtrigger is sequential by default.

Reviewer: (Sorry for butchering your text - it's much easier to edit a paragraph than to write it in the first place 😁 )

Author: No problem, you're clearly more eloquent than I.

> A trigger can be set as sequential in any/all of the following ways.

Reviewer: ```suggestion
A trigger can be set as sequential in the following ways.
```

Author: I guess I want to suggest you can use all with precedence taken into account.

Reviewer: OK I don't mind if you keep that.  Maybe say "any or all of" though.

By setting the workflow wide ``sequential xtriggers`` (defaults
to ``False``) and/or keyword argument ``sequential`` to ``True``/``False`` in
the xtrigger declaration:

.. literalinclude:: ../../workflows/xtrigger/sequential/flow.cylc
   :language: cylc

> And/or via a ``sequential`` keyword argument default in the trigger function

Reviewer: ```suggestion
And/or via a ``sequential`` keyword argument in the trigger function
```

    package.

- they can take arbitrary positional and keyword arguments

>   (``sequential`` is reserved, see :ref:`Sequential Xtriggers`)

Reviewer 2: ```suggestion
  (except ``sequential``, which is reserved - see :ref:`Sequential Xtriggers`)
```

>   (``sequential`` is reserved, see :ref:`Sequential Trigger Spawning`)

Reviewer: ```suggestion
  (``sequential`` is reserved, see :ref:`Sequential Xtriggers`)
```
